File: 2015/day01.py
# Advent of Code, 2015
# Day 1
# 11/29/2020

def what_floor(instructions): 
    '''
    Given instructions as a sequence of open or closed parentheses, e.g. "()())()()"
    return a the floor that you end up on. Open paren means go up one floor. Close
    paren means go down one floor. You start on floor 0.
    '''

    current_floor = 0
    first_visits_basement_at = 0

    for instruction_number, instruction in enumerate(instructions):
        if instruction == "(":
            current_floor = current_floor + 1
        elif instruction == ')':
            current_floor = current_floor - 1
        else:
            raise ValueError(f"Unexpected instruction encountered {instruction}")
        
        if first_visits_basement_at == 0 and current_floor < 0:
            first_visits_basement_at = instruction_number + 1

    return current_floor, first_visits_basement_at 


if __name__ == "__main__":

    # Write some tests of our floor counter function
    test_instructions = ")())())" 

    assert what_floor(test_instructions) == (-3, 1)

    with open('./inputs/day01.txt', 'r') as f:
        instructions = f.read()
    
    final_floor, first_basement_at = what_floor(instructions)
    print(f"You ended up on floor {final_floor}")
    print(f"You first went to the basement at step {first_basement_at}")

File: 2015/day01.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <string>

using namespace std;

vector<string> parse_file(string filename) {
    ifstream input_file;
    input_file.open(filename, ios::in);
    string line;
    vector<string> file_contents;

    if (input_file.is_open())
    {
        while( getline(input_file, line) ){
            file_contents.push_back(line);
        }
        input_file.close();
    }
    return file_contents;
}


int what_floor(string test_instruction) {
    int floor = 0;
    for (const char& c: test_instruction){
        switch(c) {
            case '(':
                floor++;
                break;
            case ')':
                floor--;
                break;
            default:
                break;
        }
    }
    return floor;
}


int first_in_basement_at(string test_instruction) {
    int floor = 0;
    for (int i = 0; i < test_instruction.size(); i++){
        char& c = test_instruction[i];
        switch(c) {
            case '(':
                floor++;
                break;
            case ')':
                floor--;
                break;
            default:
                break;
        }
        if (floor < 0){
            return i + 1;
        }
    }
    return 0;
}


int main()
{

    string test_instruction {"(()))("};
    int floor_num = what_floor(test_instruction);
    cout << "TESTS" << endl;
    cout << "Floor number " << floor_num << endl;
    cout << "First in basement " << first_in_basement_at(test_instruction) << endl;
    cout << endl;

    vector<string> instructions = parse_file("./inputs/day01.txt");

    // for (auto i = instructions.begin(); i != instructions.end(); ++i){
    //     cout << *i << endl << endl;
    // }

    string instruction = *instructions.begin();
    cout << "REAL THING" << endl;
    cout << "Floor number: " << what_floor(instruction) << endl;
    cout << "First in basement: " << first_in_basement_at(instruction) << endl;
}
